provider/pkg/provider/remote/copyController.go

```diff
@@ -0,0 +1,195 @@
+	var failures []p.CheckFailure
+
+	hasAsset := newInputs.HasValue("asset")
+	hasArchive := newInputs.HasValue("archive")
+
+	if hasAsset && hasArchive {
+		failures = append(failures, p.CheckFailure{
+			Property: "asset",
+			Reason:   "only one of asset or archive can be set",
+		})
+	}
+	if !hasAsset && !hasArchive {
+		failures = append(failures, p.CheckFailure{
+			Property: "asset",
+			Reason:   "either asset or archive must be set",
+		})
+	}
+
+	inputs, newFailures, err := infer.DefaultCheck[CopyInputs](newInputs)
+	if err != nil {
+		return inputs, failures, err
+	}
+	failures = append(failures, newFailures...)
+
+	if hasAsset && !inputs.Asset.IsPath() {
+		failures = append(failures, p.CheckFailure{
+			Property: "asset",
+			Reason:   "asset must be a path-based file asset",
+		})
+	}
+	if hasArchive && !inputs.Archive.IsPath() {
+		failures = append(failures, p.CheckFailure{
+			Property: "archive",
+			Reason:   "archive must be a path to a file or directory",
+		})
+	}
+
+	return inputs, failures, nil
+}
+
+func doCopy(ctx p.Context, input CopyInputs) (CopyOutputs, error) {
+	sourcePath := input.sourcePath()
+
+	ctx.Logf(diag.Debug,
+		"Creating remote: %s:%s from local '%s'",
+		input.Connection.Host, input.RemotePath, sourcePath)
+
+	client, err := input.Connection.Dial(ctx)
+	if err != nil {
+		return CopyOutputs{input}, err
+	}
+	defer client.Close()
+
+	sftp, err := sftp.NewClient(client)
+	if err != nil {
+		return CopyOutputs{input}, err
+	}
+	defer sftp.Close()
+
+	src, err := os.Open(sourcePath)
+	if err != nil {
+		return CopyOutputs{input}, err
+	}
+	defer src.Close()
+
+	srcInfo, err := src.Stat()
+	if err != nil {
+		return CopyOutputs{input}, err
+	}
+	if srcInfo.IsDir() {
+		err = copyDir(sftp, sourcePath, input.RemotePath)
+	} else {
+		err = copyFile(sftp, sourcePath, input.RemotePath)
+	}
+	return CopyOutputs{input}, err
+}
+
+// This is the Create method. This will be run on every Copy resource creation.
+func (*Copy) Create(ctx p.Context, name string, input CopyInputs, preview bool) (string, CopyOutputs, error) {
+	if preview {
+		return "", CopyOutputs{input}, nil
+	}
+
+	outputs, err := doCopy(ctx, input)
+	if err != nil {
+		return "", CopyOutputs{input}, err
+	}
+
+	id, err := resource.NewUniqueHex("", 8, 0)
+	return id, outputs, err
+}
+
+func (c *Copy) Update(ctx p.Context, id string, olds CopyOutputs, news CopyInputs, preview bool) (CopyOutputs, error) {
+	needCopy := true
+	if news.hash() == olds.hash() && news.RemotePath == olds.RemotePath {
+		needCopy = false
+	}
+
+	if preview {
+		// TODO,tkappler how to show the user whether we need to copy or not?
+		return CopyOutputs{news}, nil
+	}
+
+	if needCopy {
+		return doCopy(ctx, news)
+	}
+	return CopyOutputs{news}, nil
+}
+
+func copyFile(sftp *sftp.Client, src, dst string) error {
+	local, err := os.Open(src)
+	if err != nil {
+		return err
+	}
+	defer local.Close()
+
+	remote, err := sftp.Create(dst)
```

Comment:
What if the parents of dst do not exist?

Reply:
When `copyFile` is called from our `copyDir`, that shouldn't be possible.

When it's called directly because the user wants to copy a file, one could argue that it should fail because the user asked to copy to a non-existant location. It's one of the questions in my (internal) ideation doc.